Index: lm_sensors-3.3.5/Makefile
===================================================================
--- lm_sensors-3.3.5.orig/Makefile
+++ lm_sensors-3.3.5/Makefile
@@ -80,7 +80,7 @@ MANDIR := $(PREFIX)/man
 MACHINE := $(shell uname -m)
 
 # Extra non-default programs to build; e.g., sensord
-#PROG_EXTRA := sensord
+PROG_EXTRA := sensord
 
 # Build and install static library
 BUILD_STATIC_LIB := 1
