--- a/net/wireless/wext-compat.c
+++ b/net/wireless/wext-compat.c
@@ -1274,6 +1274,7 @@ static int cfg80211_wext_giwrate(struct
 	if (err)
 		return err;
 
+	sinfo.filled &= ~BIT(NL80211_STA_INFO_TX_BITRATE);
 	err = rdev_get_station(rdev, dev, addr, &sinfo);
 	if (err)
 		return err;
@@ -1311,8 +1312,8 @@ static struct iw_statistics *cfg80211_wi
 	memcpy(bssid, wdev->current_bss->pub.bssid, ETH_ALEN);
 	wdev_unlock(wdev);
 
-	memset(&sinfo, 0, sizeof(sinfo));
-
+	sinfo.filled &= ~BIT(NL80211_STA_INFO_SIGNAL);
+	sinfo.filled &= ~BIT(NL80211_STA_INFO_TX_FAILED);
 	if (rdev_get_station(rdev, dev, bssid, &sinfo))
 		return NULL;
 
