--- a/drivers/net/wireless/ath/ath10k/wmi-tlv.c
+++ b/drivers/net/wireless/ath/ath10k/wmi-tlv.c
@@ -1580,10 +1580,10 @@ ath10k_wmi_tlv_op_gen_start_scan(struct ath10k *ar,
 	bssid_len = arg->n_bssids * sizeof(struct wmi_mac_addr);
 	ie_len = roundup(arg->ie_len, 4);
 	len = (sizeof(*tlv) + sizeof(*cmd)) +
-	      (arg->n_channels ? sizeof(*tlv) + chan_len : 0) +
-	      (arg->n_ssids ? sizeof(*tlv) + ssid_len : 0) +
-	      (arg->n_bssids ? sizeof(*tlv) + bssid_len : 0) +
-	      (arg->ie_len ? sizeof(*tlv) + ie_len : 0);
+	      (arg->n_channels ? sizeof(*tlv) + chan_len : sizeof(*tlv)) +
+	      (arg->n_ssids ? sizeof(*tlv) + ssid_len : sizeof(*tlv)) +
+	      (arg->n_bssids ? sizeof(*tlv) + bssid_len :  sizeof(*tlv)) +
+	      (arg->ie_len ? sizeof(*tlv) + ie_len : sizeof(*tlv));
 
 	skb = ath10k_wmi_alloc_skb(ar, len);
 	if (!skb)
