diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 4701a0e..7937bfe 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1428,9 +1428,11 @@ struct ieee80211_channel_switch {
  * to indicate which TID configuration parameter changed.
  *
  * @TID_RETRY_CONF_CHANGED: retry configuration changed.
+ * @TID_AGGR_CONF_CHANGED: Aggregation config changed for the TID.
  */
 enum ieee80211_tid_conf_change {
 	TID_RETRY_CONF_CHANGED		= BIT(0),
+	TID_AGGR_CONF_CHANGED		= BIT(1),
 };
 
 /*
@@ -1440,11 +1442,13 @@ enum ieee80211_tid_conf_change {
  * @tid: TID number
  * @retry_short: retry count value
  * @retry_long: retry count value
+ * @aggr: enable/disable aggregation
  */
 struct ieee80211_tid_conf {
 	u8 tid;
 	int retry_short;
 	int retry_long;
+	bool aggr;
 };
 /**
  * enum ieee80211_vif_flags - virtual interface flags
diff --git a/net/mac80211/cfg.c b/net/mac80211/cfg.c
index f6db003..fbd9d93 100644
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -3812,6 +3812,40 @@ static int ieee80211_set_data_retry_count(struct wiphy *wiphy,
 	return ret;
 }
 
+static int ieee80211_set_tid_aggr_config(struct wiphy *wiphy,
+					 struct net_device *dev,
+					 const u8 *peer, u8 tid,
+					 bool aggr)
+{
+	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
+	struct sta_info *sta;
+	int ret;
+
+	if (!sdata->local->ops->set_tid_conf)
+		return -EOPNOTSUPP;
+
+	sdata->vif.tid_conf.tid = tid;
+	sdata->vif.tid_conf.aggr = aggr;
+
+	if (!peer)
+		return drv_set_tid_conf(sdata->local, sdata, NULL,
+					TID_AGGR_CONF_CHANGED);
+
+	mutex_lock(&sdata->local->sta_mtx);
+
+	sta = sta_info_get_bss(sdata, peer);
+	if (!sta) {
+		mutex_unlock(&sdata->local->sta_mtx);
+		return -ENOENT;
+	}
+
+	ret = drv_set_tid_conf(sdata->local, sdata, &sta->sta,
+			       TID_AGGR_CONF_CHANGED);
+
+	mutex_unlock(&sdata->local->sta_mtx);
+	return ret;
+}
+
 const struct cfg80211_ops mac80211_config_ops = {
 	.add_virtual_intf = ieee80211_add_iface,
 	.del_virtual_intf = ieee80211_del_iface,
@@ -3906,4 +3940,5 @@ const struct cfg80211_ops mac80211_config_ops = {
 	.del_nan_func = ieee80211_del_nan_func,
 	.set_multicast_to_unicast = ieee80211_set_multicast_to_unicast,
 	.set_data_retry_count = ieee80211_set_data_retry_count,
+	.set_tid_aggr_config = ieee80211_set_tid_aggr_config,
 };
