--- a/modules/admin-full/luasrc/model/cbi/admin_network/wifi.lua
+++ b/modules/admin-full/luasrc/model/cbi/admin_network/wifi.lua
@@ -163,9 +163,19 @@
 end
 
 if has_sta then
-	ch = s:taboption("general", DummyValue, "choice", translate("Channel"))
-	ch.value = translatef("Locked to channel %d used by %s",
-		has_sta:channel(), has_sta:shortname())
+	if has_sta:is_up() and has_sta:channel()~=nil then
+		ch = s:taboption("general", DummyValue, "choice", translate("Channel"))
+		ch.value = translatef("Locked to channel %d used by %s",
+			has_sta:channel(), has_sta:shortname())
+	else
+		ch = s:taboption("general", Value, "channel", translate("Channel"))
+		ch:value("auto", translate("auto"))
+		for _, f in ipairs(iw and iw.freqlist or { }) do
+			if not f.restricted then
+				ch:value(f.channel, "%i (%.3f GHz)" %{ f.channel, f.mhz / 1000 })
+			end
+		end
+	end
 else
 	ch = s:taboption("general", Value, "channel", translate("Channel"))
 	ch:value("auto", translate("auto"))
