Index: musl-1.1.16/src/stdio/fputc.c
===================================================================
--- musl-1.1.16.orig/src/stdio/fputc.c	2017-01-01 08:57:17.000000000 +0530
+++ musl-1.1.16/src/stdio/fputc.c	2017-10-20 12:10:54.073719904 +0530
@@ -1,5 +1,15 @@
 #include "stdio_impl.h"
+#define UNBUFFERED_PUTC
 
+#ifdef UNBUFFERED_PUTC
+int fputc(int c, FILE *f)
+{
+	FLOCK(f);
+	f->write(f, &c, 1);
+	FUNLOCK(f);
+	return c;
+}
+#else
 int fputc(int c, FILE *f)
 {
 	if (f->lock < 0 || !__lockfile(f))
@@ -8,3 +18,4 @@
 	__unlockfile(f);
 	return c;
 }
+#endif
