Index: libnl-3.2.21/configure.ac
===================================================================
--- libnl-3.2.21.orig/configure.ac	2013-01-24 19:06:06.000000000 +0530
+++ libnl-3.2.21/configure.ac	2017-04-12 15:12:42.960832043 +0530
@@ -95,7 +95,7 @@
 if test "x$enable_pthreads" = "xno"; then
     AC_DEFINE([DISABLE_PTHREADS], [1], [Define to 1 to disable pthreads])
 else
-    AC_CHECK_LIB([pthread], [pthread_mutex_lock], [], AC_MSG_ERROR([libpthread is required]))
+    AC_CHECK_LIB([pthread], [pthread_mutex_lock], [], AC_MSG_ERROR([libpthread is required]), [-ldl])
 fi
 
 AC_CONFIG_SUBDIRS([doc])
